proposals/0000-implicit-resource-binding.md

```diff
@@ -0,0 +1,348 @@
+RWBuffer<float> A : register(u1);
+RWBuffer<float> D[4];    // gets u6
+RWBuffer<float> B[];     // gets u10 (unbounded range)
+RWBuffer<float> C : register(u5);
+
+[numthreads(4,1,1)]
+void main() {
+  A[0] = D[2][0] + C[0] + B[10][0];
+}
+```
+https://godbolt.org/z/cjP35n63d
+
+Since DXC seems to only assigns descriptors from `space0`, it reports error
+whenever there are:
+- two or more unbound resource arrays without explicit binding
+- one unboud resource array with explicit binding in `space0` and one or more
+  unbound resource arrays without explicit binding
+
+#### Example 3.4
+
+```c++
+RWBuffer<float> C[];  // gets u0 (unbounded range)
+RWBuffer<float> D[];  // error
+
+[numthreads(4,1,1)]
+void main() {
+  D[10][0] = C[20][2];
+}
+```
+https://godbolt.org/z/od7o6xfWE
+
+Unbound resource arrays cannot be placed inside structs and DXC reports `error:
+array dimensions of struct/class members must be explicit`.
+
+### Register spaces other than `space0`
+
+Use of other virtual register spaces does not seem to affect the implicit
+assignments in `space0`:
+
+#### Example 4.1
+```c++
+....
+RWBuffer<float> A : register(u0, space1); 
+RWBuffer<float> B[10];   // gets u0 in space0 (range 10)
+RWBuffer<float> C[5][4] : register(u10, space5);
+RWBuffer<float> D;       // gets u10 in space0
+
+[numthreads(4,1,1)]
+void main() {
+  A[0] = D[0] + C[3][3][0] + B[5][0];
+}
+```
+https://godbolt.org/z/EeW17MncE
+
+### Unused resources
+
+Unused resources are resources that are not referenced in the source code or
+that are unreachable from the shader entry point function and later removed from
+the final IR code by LLVM optimization passes.
+
+When a resource is unused and it has an explicit binding, this binding is
+ignored and its descriptor range is available for use by other resources.
+
+In the following example the resources `B`, `C` and `s.D` are identified as
+unused and optimized away. Resources `A` and `E` get the binding from descriptor
+range originally assigned to `B`:
+
+#### Example 5.1
+
+```c++
+struct S {
+  RWBuffer<float> D;    
+};
+
+RWBuffer<float> A; // gets u0
+RWBuffer<float> B[10] : register(u0);  // unused
+RWBuffer<float> C : register(u2);      // unused
+S s : register(u5);                    // unused
+RWBuffer<float> E[5]; // gets u1
+
+void foo() {
+    B[2][0] = 1.0;
+}
+
+bool bar(int i) {
+    return i > 10;
+}
+
+[numthreads(4,1,1)]
+void main(uint3 gi : SV_GROUPID) {
+  if (false) {
+    C[0] = 1.0;
+  }
+  for (int i = 0; i < 5; i++) {
+    if (i > 6)
+      s.D[0] = 1.0; // unreachable
+    if (bar(i))
+      foo();        // unreachable
+    if (bar(gi.x))
+      A[0] = E[2][i];
+  }
+}
+```
+https://godbolt.org/z/Mcabxq4jG
+
+Based on resources identified as unused in the example above it is very likely
+that DXC is assigning implicit bindings later in the compiler after codegen and
+code optimizations.  
+
+## Motivation
+
+We need to support implicit resource bindings in Clang as it is a fundamental
```

Comment:
I will probably lose this argument, but I would challenge this statement. Implicit bindings are not allowed under certain scenarios, and make various tasks like GBV (gpu based validation) more difficult to debug as the implementation must be kept in the loop. Rather than being able to validate the root signature offline.

Yes, we can have a container that doesn't have a root signature, but, in the absence of one, the app needs to create one. If the app creates one, and they mismatch with one that's in a container, the PSO creation fails.

In my opinion, this is a shortcoming of DXC, not behavior we want to bring forward.

Reply:
I'm not sure I follow the logic here. We can have implicit bindings and a root signature. We can also have no implicit bindings and have a runtime-provided root signature. For example this is completely valid:

```hlsl
cbuffer RootConstants {
  float4 C;
};

struct Input {
  float4 A;
  float4 B;
};

struct Output {
  float4 A;
};

StructuredBuffer<Input> In;
RWStructuredBuffer<Output> Out;

[RootSignature("CBV(b0), SRV(t0), UAV(u0)")]
[numthreads(1,1,1)]
void main(uint GI : SV_GroupIndex) {
  Out[GI].A = In[GI].A * In[GI].B * C;
}
```

Are you arguing that this should not be valid?

Reply:
Before I started working on the compiler I'd never considered embedding a root signature in a shader and always constructed them in my C++ code.  As far as I was concerned, my C++ code was specifying the interface - the root signature - that the shaders I used had to comply with.  Although there are reasons that this workflow isn't the most efficient, I don't think we should prohibit it.

Reply:
I think we're also talking about different things.

Whether or not a root signature is provided in the shader, and whether or not resources are assigned register values implicitly are completely different.

At the DXIL level you can't even tell if a shader's bindings are implicit or explicit. Consider for example this shader:
```
#if IMPLICIT
#define BINDING(val)
#else 
#define BINDING(val) : register(val)
#endif

RWBuffer<int4> Buf BINDING(u0);

[numthreads(1,1,1)]
void main() {
    Buf[0] += Buf[0];
}
```
[Compiler Explorer](https://godbolt.org/z/WKe5bes4o)

Whether you compile this with `-DIMPLICIT` or not, the output is identical.

Reply:
I think I have to do some more reading as I'm confusing myself. Sorry, please disregard.